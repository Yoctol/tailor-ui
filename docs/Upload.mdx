---
name: Upload
menu: Components
route: /components/upload
---

import { PropsTable, Playground } from 'docz';
import { Modal, Upload } from 'tailor-ui';

# Upload

Upload file by selecting or dragging.

## When To Use

- When you need to upload one or more files.
- When you need to show the process of uploading.
- When you need to upload files by dragging and dropping.

## Examples

```js
import { Upload } from 'tailor-ui';
```

### Basic

<Playground>
  <Upload onSelect={() => new Promise(resolve => setTimeout(resolve, 2000))} />
</Playground>

### before select

<Playground>
  <Upload
    onBeforeSelect={() =>
      Modal.confirm({
        title: 'Do you want to override items?',
        content: 'It can not fallback',
      })
    }
    onSelect={() => new Promise(resolve => setTimeout(resolve, 2000))}
  />
</Playground>

### Multiple

<Playground>
  <Upload
    multiple
    onSelect={() => new Promise(resolve => setTimeout(resolve, 2000))}
  />
</Playground>

### failed and success

<Playground>
  {() => {
    const [uploadFirstTime, setUploadFirstTime] = React.useState(true);

    return (
      <Upload
        multiple
        onSelect={() => new Promise((resolve, reject) => {
          if (uploadFirstTime) {
            setTimeout(reject, 1000);
            setUploadFirstTime(false);
          } else {
            setTimeout(resolve, 1000);
          }
        })}
      />
    );

}}

</Playground>

## API

This component is based on [react-dropzone](https://github.com/react-dropzone/react-dropzone), please check [it's documentation](https://react-dropzone.js.org/#proptypes) for more usage.

<PropsTable of={Upload} />
