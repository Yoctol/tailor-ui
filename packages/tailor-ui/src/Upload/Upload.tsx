import React, {
  FunctionComponent,
  Reducer,
  useContext,
  useReducer,
} from 'react';
import styled, { css } from 'styled-components';
import { DropzoneProps, useDropzone } from 'react-dropzone';
import { MdCheck, MdClose, MdFileUpload } from 'react-icons/md';

import Box from '../Grid/Box';
import Button from '../Button';
import Flex from '../Grid/Flex';
import Icon from '../Icon';
import { LocaleContext } from '../UIProvider';

const FileList = styled.div`
  display: inline-flex;
  flex-direction: column;
`;

const getUploadText = ({
  uploading,
  uploaded,
  failed,
  texts,
}: {
  uploading: boolean;
  uploaded: boolean;
  failed: boolean;
  texts: {
    uploadText: string;
    uploadingText: string;
    uploadedText: string;
    failedText: string;
  };
}) => {
  if (uploading) {
    return texts.uploadingText;
  }

  if (failed) {
    return texts.failedText;
  }

  if (uploaded) {
    return texts.uploadedText;
  }

  return texts.uploadText;
};

const getUploadIcon = ({
  uploaded,
  failed,
}: {
  uploaded: boolean;
  failed: boolean;
}) => {
  if (uploaded) {
    return MdCheck;
  }

  if (failed) {
    return MdClose;
  }

  return MdFileUpload;
};

interface FileItemProps {
  file: File;
  uploaded: boolean;
  onClear: (file: File) => void;
}

const FileItem: FunctionComponent<FileItemProps> = ({
  file,
  uploaded,
  onClear,
}) => (
  <Flex
    alignItems="center"
    mt="1"
    p="1"
    borderRadius="base"
    css={css`
      &:hover {
        background-color: ${p => p.theme.colors.gray200};
      }
    `}
  >
    {uploaded && (
      <Icon
        type={MdClose}
        cursor="pointer"
        mr="1"
        size="16"
        onClick={() => onClear(file)}
      />
    )}
    <Box color="gray600" fontSize="sm">
      {file.name}
    </Box>
  </Flex>
);

interface UploadProps extends DropzoneProps {
  onSelect: (files: File[]) => Promise<any>;
  onClear?: (file: File) => void;
  onBeforeSelect?: () => Promise<boolean> | boolean;
  texts?: {
    uploadText?: string;
    uploadingText?: string;
    uploadedText?: string;
    failedText?: string;
  };
}

const START_UPLOADING = 'START_UPLOADING';
const UPLOAD_SUCCESS = 'UPLOAD_SUCCESS';
const UPLOAD_FAILED = 'UPLOAD_FAILED';
const CLEAR_FILES = 'CLEAR_FILES';

interface ReducerAction {
  type: 'START_UPLOADING' | 'UPLOAD_SUCCESS' | 'UPLOAD_FAILED' | 'CLEAR_FILES';
  payload?: any;
}

interface ReducerState {
  uploading: boolean;
  uploaded: boolean;
  failed: boolean;
  files: File[];
}

const uploadReducer: Reducer<ReducerState, ReducerAction> = (state, action) => {
  switch (action.type) {
    case START_UPLOADING:
      return {
        ...state,
        uploading: true,
        uploaded: false,
        failed: false,
        files: action.payload,
      };
    case UPLOAD_SUCCESS:
      return {
        ...state,
        uploading: false,
        uploaded: true,
        failed: false,
      };
    case UPLOAD_FAILED:
      return {
        ...state,
        uploading: false,
        uploaded: false,
        failed: true,
      };
    case CLEAR_FILES:
      return {
        ...state,
        uploaded: false,
        files: state.files.filter(file => file.name !== action.payload),
      };
    default:
      throw new Error();
  }
};

const Upload: FunctionComponent<UploadProps> = ({
  onBeforeSelect,
  onSelect,
  onClear,
  disabled,
  texts = {},
  ...props
}) => {
  const { locale } = useContext(LocaleContext);
  const [state, dispatch] = useReducer(uploadReducer, {
    uploading: false,
    uploaded: false,
    failed: false,
    files: [],
  });

  const handleSelect = async (selectedFiles: File[]) => {
    dispatch({
      type: 'START_UPLOADING',
      payload: selectedFiles,
    });

    try {
      await onSelect(selectedFiles);
      dispatch({ type: 'UPLOAD_SUCCESS' });
    } catch {
      dispatch({ type: 'UPLOAD_FAILED' });
    }
  };

  const handleClear = (clearFile: File) => {
    dispatch({ type: 'CLEAR_FILES', payload: clearFile.name });

    if (onClear) {
      onClear(clearFile);
    }
  };

  const icon = getUploadIcon(state);
  const text = getUploadText({
    ...state,
    texts: {
      ...locale.Upload,
      ...texts,
    },
  });
  const { getRootProps, getInputProps, open } = useDropzone({
    onDrop: handleSelect,
  });

  return (
    <div>
      <div
        tabIndex={-1}
        onKeyPress={() => {}}
        role="button"
        {...getRootProps()}
        style={{ display: 'inline-flex' }}
        onClick={event => event.preventDefault()}
      >
        <input {...getInputProps()} />
        <Button
          icon={icon}
          loading={state.uploading}
          disabled={disabled}
          onClick={async event => {
            event.preventDefault();

            if (onBeforeSelect) {
              try {
                const success = await onBeforeSelect();
                if (!success) {
                  return;
                }
              } catch (error) {
                console.error(error);

                return;
              }
            }

            open();
          }}
          {...props}
        >
          {text}
        </Button>
      </div>
      {state.files.length > 0 && (
        <FileList>
          {state.files.map(file => (
            <FileItem
              key={file.name}
              file={file}
              uploaded={state.uploaded}
              onClear={handleClear}
            />
          ))}
        </FileList>
      )}
    </div>
  );
};

export default Upload;
