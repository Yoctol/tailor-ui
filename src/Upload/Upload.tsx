import Dropzone, { DropzoneProps } from 'react-dropzone';
import React, { FunctionComponent, useContext, useState } from 'react';
import { MdCheck, MdClose, MdFileUpload } from 'react-icons/md';

import styled, { css } from 'utils/styled-components';

import Box from '../Grid/Box';
import Button from '../Button';
import Flex from '../Grid/Flex';
import Icon from '../Icon';
import { LocaleContext, LocaleType } from '../UIProvider';

type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

const FileList = styled.div`
  display: inline-flex;
  flex-direction: column;
`;

const getUploadText = ({
  uploading,
  uploaded,
  failure,
  locale,
}: {
  uploading: boolean;
  uploaded: boolean;
  failure: boolean;
  locale: LocaleType;
}) => {
  const texts = locale.Upload;

  if (uploading) {
    return texts.uploadingText;
  }

  if (failure) {
    return texts.failureText;
  }

  if (uploaded) {
    return texts.uploadedText;
  }

  return texts.uploadText;
};

const getUploadIcon = ({
  uploaded,
  failure,
}: {
  uploaded: boolean;
  failure: boolean;
}) => {
  if (uploaded) {
    return MdCheck;
  }

  if (failure) {
    return MdClose;
  }

  return MdFileUpload;
};

interface IFileItemProps {
  file: File;
  onClear: (file: File) => void;
}

const FileItem: FunctionComponent<IFileItemProps> = ({ file, onClear }) => (
  <Flex
    alignItems="center"
    mt="1"
    p="1"
    borderRadius="base"
    css={css`
      &:hover {
        background-color: ${p => p.theme.colors.gray200};
      }
    `}
  >
    <Icon
      type={MdClose}
      cursor="pointer"
      mr="1"
      size="16"
      onClick={() => onClear(file)}
    />
    <Box color="gray600" fontSize="sm">
      {file.name}
    </Box>
  </Flex>
);

interface IUploadProps extends Omit<DropzoneProps, 'ref' | 'onSelect'> {
  onSelect: (files: File[]) => Promise<any>;
  onClear?: (file: File) => void;
}

const Upload: FunctionComponent<IUploadProps> = ({
  onSelect,
  onClear,
  disabled,
  ...props
}) => {
  const { locale } = useContext(LocaleContext);
  const [uploading, setUploading] = useState(false);
  const [uploaded, setUploaded] = useState(false);
  const [failure, setFailure] = useState(false);
  const [files, setFiles] = useState<File[]>([]);

  const handleSelect = async (selectedFiles: File[]) => {
    setFiles(selectedFiles);
    setUploading(true);

    try {
      await onSelect(selectedFiles);
      setUploaded(true);
    } catch {
      setFailure(true);
    }

    setUploading(false);
  };

  const handleClear = (clearFile: File) => {
    setUploaded(false);
    setFiles(prevFiles =>
      prevFiles.filter(file => file.name !== clearFile.name)
    );

    if (onClear) {
      onClear(clearFile);
    }
  };

  const text = getUploadText({ uploading, uploaded, failure, locale });
  const icon = getUploadIcon({ uploaded, failure });

  return (
    <div>
      <Dropzone
        disableClick
        style={{ position: 'relative' }}
        onDrop={handleSelect}
        {...props}
      >
        {({ open }) => (
          <Button
            icon={icon}
            loading={uploading}
            disabled={disabled}
            onClick={() => {
              if (uploaded) {
                setUploaded(false);
              }
              open();
            }}
          >
            {text}
          </Button>
        )}
      </Dropzone>
      <FileList>
        {files.map(file => (
          <FileItem key={file.name} file={file} onClear={handleClear} />
        ))}
      </FileList>
    </div>
  );
};

export default Upload;
