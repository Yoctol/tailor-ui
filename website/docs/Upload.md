---
id: upload
title: Upload
---

Upload file by selecting or dragging.

## When To Use

- When you need to upload one or more files.
- When you need to show the process of uploading.
- When you need to upload files by dragging and dropping.

## Examples

```js
import { Upload } from 'tailor-ui';
```

### Basic

```jsx live
<Upload onSelect={() => new Promise((resolve) => setTimeout(resolve, 2000))} />
```

### before select

```jsx live
() => {
  const modal = useModal();

  return (
    <Upload
      onBeforeSelect={() =>
        modal.confirm({
          title: 'Do you want to override items?',
          content: 'It can not fallback',
        })
      }
      onSelect={() => new Promise((resolve) => setTimeout(resolve, 2000))}
    />
  );
};
```

### Multiple

```jsx live
<Upload
  multiple
  onSelect={() => new Promise((resolve) => setTimeout(resolve, 2000))}
/>
```

### failed and success

```jsx live
() => {
  const [uploadFirstTime, setUploadFirstTime] = useState(true);

  return (
    <Upload
      multiple
      onSelect={() =>
        new Promise((resolve, reject) => {
          if (uploadFirstTime) {
            setTimeout(reject, 1000);
            setUploadFirstTime(false);
          } else {
            setTimeout(resolve, 1000);
          }
        })
      }
    />
  );
};
```

## API

| Property         | Description | Type                                                                                           | Default |
| ---------------- | ----------- | ---------------------------------------------------------------------------------------------- | ------- |
| `disabled`       |             | `boolean`                                                                                      |         |
| `onSelect`       |             | `(files: File[]) => Promise<any>`                                                              |         |
| `onClear`        |             | `(file: File) => void`                                                                         |         |
| `onBeforeSelect` |             | `() => Promise<boolean>` \| `boolean`                                                          |         |
| `buttonProps`    |             | `ButtonProps`                                                                                  |         |
| `texts`          |             | `{ uploadText?: string; uploadingText?: string; uploadedText?: string; failedText?: string; }` |         |

This component is based on [react-dropzone](https://github.com/react-dropzone/react-dropzone), please check [it's documentation](https://react-dropzone.js.org/#proptypes) for more usage.
