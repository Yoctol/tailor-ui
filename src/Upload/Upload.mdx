---
name: Upload
menu: Components
route: /components/upload
---

import { PropsTable, Playground } from 'docz';

import Upload from './';

# Upload

Upload file by selecting or dragging.

## When To Use

- When you need to upload one or more files.
- When you need to show the process of uploading.
- When you need to upload files by dragging and dropping.

## Examples

```js
import { Upload } from 'tailor-ui';
```

### Basic

<Playground>
  <Upload onSelect={() => new Promise(resolve => setTimeout(resolve, 2000))} />
</Playground>

### Multiple

<Playground>
  <Upload
    multiple
    onSelect={() => new Promise(resolve => setTimeout(resolve, 2000))}
  />
</Playground>

## API

This component is based on [react-dropzone](https://github.com/react-dropzone/react-dropzone), please check [it's documentation](https://react-dropzone.js.org/#proptypes) for more usage.

<PropsTable of={Upload} />
