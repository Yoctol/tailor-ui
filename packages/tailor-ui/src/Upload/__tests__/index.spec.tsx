import React from 'react';

import { fireEvent, render, waitForElement } from 'test/test-utils';

import Upload from '../Upload';

const setup = async ({
  handleSelect,
  handleClear,
  expectFinalStateText,
}: {
  handleSelect: () => Promise<any>;
  handleClear?: () => void;
  expectFinalStateText: string;
}) => {
  const renderer = render(
    <Upload onSelect={handleSelect} onClear={handleClear} />
  );

  const { container, getByText } = renderer;

  const uploadInput = container.querySelector('input[type=file]');

  const file = new File(['example'], 'example.json', {
    type: 'application/json',
  });

  fireEvent.click(getByText('Click to Upload'));
  fireEvent.change(uploadInput, { target: { files: [file] } });

  // FIXME: Can not get the Uploading... button
  // await waitForElement(() => getByText('Uploading...'));
  await waitForElement(() => getByText('example.json'));
  expect(container).toContainElement(getByText('example.json'));

  await waitForElement(() => getByText(expectFinalStateText));

  expect(container).toHaveTextContent(expectFinalStateText);

  return {
    ...renderer,
    file,
  };
};

describe('Upload', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Upload onSelect={files => new Promise(resolve => resolve(files))} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should show correctly text when uploaded success', async () => {
    const handleSelect = jest.fn().mockResolvedValue(true);
    const expectFinalStateText = 'Upload Completed!';

    await setup({ handleSelect, expectFinalStateText });
  });

  it('should show correctly text when uploaded failed', async () => {
    const handleSelect = jest.fn().mockRejectedValue(true);
    const expectFinalStateText = 'Upload Failed!';

    await setup({ handleSelect, expectFinalStateText });
  });

  it('should not show filename when clear file', async () => {
    const handleSelect = jest.fn().mockResolvedValue(true);
    const handleClear = jest.fn();
    const expectFinalStateText = 'Upload Completed!';

    const { container, queryByText, getByText, file } = await setup({
      handleSelect,
      handleClear,
      expectFinalStateText,
    });

    const clearIcon = container.querySelector('i[cursor=pointer]');

    fireEvent.click(clearIcon);

    expect(container).toContainElement(getByText('Click to Upload'));
    expect(container).not.toContainElement(queryByText('example.json'));
    expect(handleClear).toBeCalledWith(file);
  });
});
