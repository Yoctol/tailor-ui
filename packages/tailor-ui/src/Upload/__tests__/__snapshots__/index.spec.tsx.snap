// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload should render correctly 1`] = `
.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: default;
  margin-right: 0.25rem;
}

.c2 svg {
  fill: #8b8b8b;
  vertical-align: middle;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c2.c2.c2 svg {
  width: 20px;
  height: 20px;
}

.c3 {
  position: absolute;
  border-radius: 50%;
  opacity: 1;
  background: rgba(0,0,0,0.2);
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  overflow: hidden;
  border: 1px solid;
  border-radius: 0.25rem;
  line-height: 1.5;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  height: 2rem;
  padding: 0 1rem;
  font-size: 1rem;
  border-color: #e0e0e0;
  background-color: #ffffff;
  color: #1a1a1a;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c0:focus {
  outline: 0;
}

.c0:disabled,
.c0[disabled] {
  opacity: 0.7;
  cursor: default;
  pointer-events: none;
}

.c0::before {
  content: '';
  display: inline-block;
  position: relative;
  left: 0;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  width: 0;
  height: 0;
  margin-right: 0;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c0 .c1 {
  cursor: pointer;
}

.c0 .c1 svg {
  fill: #1a1a1a;
}

.c0:hover {
  border-color: #3c5ad0;
  background-color: #f6f6f6;
}

<div>
  <div
    role="button"
    style="display: inline-flex;"
    tabindex="0"
  >
    <input
      autocomplete="off"
      multiple=""
      style="display: none;"
      tabindex="-1"
      type="file"
    />
    <button
      class="c0"
      type="default"
    >
      <i
        class="c1 c2"
        cursor="default"
        style="pointer-events: none;"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z"
          />
        </svg>
      </i>
      Click to Upload
      <div
        class="c3"
      />
    </button>
  </div>
  <br />
</div>
`;
